File: lib/pearClient.ts
import axios, { AxiosInstance, AxiosError } from 'axios'
import {
  authenticateWithPrivateKey,
  getWalletAddress,
  type AuthTokens,
} from './pear-auth'

export interface BasketTradeParams {
  long: string[]  // Array of coin symbols (e.g., ['AI', 'ML'])
  short: string[] // Array of coin symbols (e.g., ['ETH'])
}

export interface BasketAsset {
  coin: string
  weight: number
}

export interface ExecuteBasketTradeResponse {
  orderId: string
  status: string
  message?: string
}

/**
 * Pear Protocol Builder Address
 * All trades are routed to this address and it must be approved before trading
 * See: https://docs.pearprotocol.io/api-integration/access-management/builder-code
 */
export const PEAR_BUILDER_ADDRESS = '0xA47D4d99191db54A4829cdf3de2417E527c3b042'

/**
 * Pear Protocol API Client
 * Real implementation for Pear Protocol API basket trade execution
 * 
 * ✅ Ported from working Python implementation (test/pear_client.py)
 * - Uses APITRADER client ID (matches Python version)
 * - Automatic authentication before trades
 * - Real JWT token-based authentication
 */
export class PearClient {
  private client: AxiosInstance
  private clientId: string
  private apiUrl: string
  private accessToken: string | null = null
  private refreshToken: string | null = null
  private privateKey: string | null = null
  private walletAddress: string | null = null

  constructor() {
    // Use APITRADER like the working Python version
    this.clientId = process.env.PEAR_CLIENT_ID || 'APITRADER'
    this.apiUrl = process.env.NEXT_PUBLIC_PEAR_API_URL || 'https://hl-v2.pearprotocol.io'
    
    // Get private key from environment (matches Python: WALLET_PRIVATE_KEY)
    this.privateKey = process.env.PEAR_WALLET_PRIVATE_KEY || null
    if (this.privateKey) {
      // Remove quotes if present
      if ((this.privateKey.startsWith('"') && this.privateKey.endsWith('"')) ||
          (this.privateKey.startsWith("'") && this.privateKey.endsWith("'"))) {
        this.privateKey = this.privateKey.slice(1, -1)
      }
      this.walletAddress = getWalletAddress(this.privateKey)
      console.log(`Pear Client initialized with wallet: ${this.walletAddress}`)
    }

    console.log(`Pear Client initialized with API URL: ${this.apiUrl}`)
    console.log(`Client ID: ${this.clientId}`)

    this.client = axios.create({
      baseURL: this.apiUrl,
      headers: {
        'Content-Type': 'application/json',
        'X-Client-Id': this.clientId,
        // Add Bearer token if we have one
        ...(this.accessToken && { 'Authorization': `Bearer ${this.accessToken}` }),
      },
    })
  }

  /**
   * Parse trade response to extract order ID and status
   */
  private _parseTradeResponse(responseData: any): ExecuteBasketTradeResponse {
    // Handle different possible response formats
    let orderId: string
    let status: string

    if (responseData.orderId) {
      orderId = responseData.orderId
    } else if (responseData.id) {
      orderId = responseData.id
    } else if (responseData.order_id) {
      orderId = responseData.order_id
    } else {
      // Fallback: use transaction hash or any identifier
      orderId = responseData.txHash || responseData.hash || 'unknown'
    }

    status = responseData.status || responseData.state || 'submitted'

    return {
      orderId,
      status,
      message: responseData.message || 'Trade executed successfully',
    }
  }

  /**
   * Set access token for authenticated requests
   */
  setAccessToken(token: string, refreshToken?: string): void {
    this.accessToken = token
    if (refreshToken) {
      this.refreshToken = refreshToken
    }
    this.client.defaults.headers.common['Authorization'] = `Bearer ${token}`
  }

  /**
   * Ensure we have a valid access token
   * If no token or expired, authenticate using private key
   */
  private async ensureAuthenticated(): Promise<void> {
    // If we have a token, assume it's valid (in production, check expiration)
    if (this.accessToken) {
      return
    }

    // Need to authenticate
    if (!this.privateKey) {
      throw new Error(
        'PEAR_WALLET_PRIVATE_KEY is required in .env.local for authentication'
      )
    }

    if (!this.walletAddress) {
      throw new Error('Wallet address could not be derived from private key')
    }

    console.log('No access token found. Authenticating...')
    const tokens = await authenticateWithPrivateKey(
      this.walletAddress,
      this.privateKey,
      this.clientId
    )

    this.setAccessToken(tokens.accessToken, tokens.refreshToken)
    console.log('✅ Authentication successful, token set')
  }

  /**
   * Check agent wallet status
   * Returns agent wallet info if it exists
   */
  async checkAgentWallet(): Promise<any> {
    await this.ensureAuthenticated()
    try {
      const response = await this.client.get('/agentWallet')
      const data = response.data
      console.log('Raw agent wallet response:', JSON.stringify(data, null, 2))
      
      // If response is empty object, the wallet exists but may need approval
      if (data && Object.keys(data).length === 0) {
        console.warn('⚠️ Agent wallet exists but appears empty/unapproved')
        console.warn('⚠️ You may need to approve it. Check if there\'s an approval endpoint.')
      }
      
      return data
    } catch (error: any) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 404) {
          console.warn('⚠️ Agent wallet not found. You may need to create one.')
          return null
        }
        console.error('Error checking agent wallet:', {
          status: error.response?.status,
          data: error.response?.data,
        })
      }
      throw error
    }
  }

  /**
   * Get agent wallet approval message (if wallet exists but needs approval)
   * This might be a separate endpoint or part of the wallet creation flow
   */
  async getAgentWalletApprovalMessage(): Promise<any> {
    await this.ensureAuthenticated()
    try {
      // Try common approval message endpoints
      const endpoints = [
        '/agentWallet/approval',
        '/agentWallet/approval-message',
        '/agentWallet/message',
      ]
      
      for (const endpoint of endpoints) {
        try {
          const response = await this.client.get(endpoint)
          console.log(`Found approval message at ${endpoint}:`, JSON.stringify(response.data, null, 2))
          return response.data
        } catch (e: any) {
          // Continue to next endpoint
          if (axios.isAxiosError(e) && e.response?.status !== 404) {
            console.warn(`Error checking ${endpoint}:`, e.response?.status)
          }
        }
      }
      
      console.warn('⚠️ Could not find approval message endpoint. Wallet may need manual approval.')
      return null
    } catch (error: any) {
      console.error('Error getting approval message:', error)
      return null
    }
  }

  /**
   * Create agent wallet
   * Returns the approval message that needs to be signed
   */
  async createAgentWallet(): Promise<any> {
    await this.ensureAuthenticated()
    
    // First check if wallet already exists
    const existingWallet = await this.checkAgentWallet()
    if (existingWallet !== null && existingWallet !== undefined) {
      const isEmpty = Object.keys(existingWallet || {}).length === 0
      if (!isEmpty) {
        console.log('Agent wallet already exists. It may need approval.')
        return { exists: true, wallet: existingWallet, needsApproval: true }
      }
    }
    
    try {
      console.log('Creating agent wallet...')
      // POST /agentWallet - The API rejects empty body with application/json content-type
      // Use a raw axios request without the default Content-Type header
      const response = await axios.post(
        `${this.apiUrl}/agentWallet`,
        undefined, // No body
        {
          headers: {
            'Authorization': `Bearer ${this.accessToken}`,
            'X-Client-Id': this.clientId,
            // Don't set Content-Type - let server handle it without JSON header
            'Content-Type': undefined,
          },
          transformRequest: [], // Disable default JSON transformation
        }
      )
      const data = response.data
      console.log('Agent wallet creation response:', JSON.stringify(data, null, 2))
      return data
    } catch (error: any) {
      if (axios.isAxiosError(error)) {
        const errorDetails = {
          status: error.response?.status,
          statusText: error.response?.statusText,
          data: error.response?.data,
          url: error.config?.url,
          method: error.config?.method,
        }
        console.error('Error creating agent wallet:', JSON.stringify(errorDetails, null, 2))
        
        // If 400, provide more helpful error message with full details
        if (error.response?.status === 400) {
          const errorData = error.response?.data
          const errorMsg = typeof errorData === 'string' 
            ? errorData 
            : JSON.stringify(errorData)
          throw new Error(`Agent wallet creation failed (400): ${errorMsg}`)
        }
      }
      throw error
    }
  }

  /**
   * Execute a basket trade through Pear Protocol
   * @param basket - Object with long and short asset arrays
   * @param notional - Notional value in USD (default: $10)
   * @returns Order ID and status from Pear Protocol
   */
  async executeBasketTrade(
    basket: BasketTradeParams,
    notional: number = 10
  ): Promise<ExecuteBasketTradeResponse> {
    // Ensure we have a valid access token before making authenticated requests
    await this.ensureAuthenticated()
    try {
      // Validate basket
      // According to Pear Protocol API docs, one of longAssets or shortAssets can be empty
      if ((!basket.long || basket.long.length === 0) && (!basket.short || basket.short.length === 0)) {
        throw new Error('At least one of long or short assets must be provided')
      }

      // Normalize coin symbols (remove /USD, /USDC, etc.)
      const normalizeCoin = (symbol: string): string => {
        return symbol.split('/')[0].trim().toUpperCase()
      }

      // Build long assets with equal weights if not specified
      const longAssets: BasketAsset[] = (basket.long || []).map((coin, index) => {
        const normalizedCoin = normalizeCoin(coin)
        // Distribute weights equally
        const weight = 1.0 / (basket.long?.length || 1)
        return { coin: normalizedCoin, weight }
      })

      // Build short assets with equal weights
      const shortAssets: BasketAsset[] = (basket.short || []).map((coin, index) => {
        const normalizedCoin = normalizeCoin(coin)
        // Distribute weights equally
        const weight = 1.0 / (basket.short?.length || 1)
        return { coin: normalizedCoin, weight }
      })

      // Verify weights sum to 1.0 (approximately, due to floating point)
      // Only check if the array is not empty
      if (longAssets.length > 0) {
        const longWeightSum = longAssets.reduce((sum, asset) => sum + asset.weight, 0)
        if (Math.abs(longWeightSum - 1.0) > 0.01) {
          throw new Error(`Long assets weights must sum to 1.0, got ${longWeightSum}`)
        }
      }
      if (shortAssets.length > 0) {
        const shortWeightSum = shortAssets.reduce((sum, asset) => sum + asset.weight, 0)
        if (Math.abs(shortWeightSum - 1.0) > 0.01) {
          throw new Error(`Short assets weights must sum to 1.0, got ${shortWeightSum}`)
        }
      }

      // Build request payload according to Pear Protocol API specification
      // POST /positions expects: executionType, leverage, usdValue, slippage, longAssets, shortAssets
      // Note: Builder code (0xA47D4d99191db54A4829cdf3de2417E527c3b042) must be approved before trades
      // See: https://docs.pearprotocol.io/api-integration/access-management/builder-code
      
      // Convert assets to use 'asset' field instead of 'coin' (API requirement)
      const longAssetsFormatted = longAssets.map(({ coin, weight }) => ({
        asset: coin, // Use 'asset' instead of 'coin'
        weight,
      }))
      
      const shortAssetsFormatted = shortAssets.map(({ coin, weight }) => ({
        asset: coin, // Use 'asset' instead of 'coin'
        weight,
      }))

      // Check agent wallet before trading (helps diagnose 500 errors)
      const agentWallet = await this.checkAgentWallet()
      const isEmptyWallet = !agentWallet || Object.keys(agentWallet).length === 0
      
      if (isEmptyWallet) {
        console.warn('⚠️ Agent wallet is empty or not found. This may cause 500 errors.')
        console.warn('⚠️ You may need to:')
        console.warn('   1. Create an agent wallet via POST /agentWallet')
        console.warn('   2. Approve the agent wallet with your wallet signature')
        console.warn('   3. Approve the builder code (0xA47D4d99191db54A4829cdf3de2417E527c3b042)')
      } else {
        console.log('Agent wallet details:', JSON.stringify(agentWallet, null, 2))
        const status = agentWallet.status || agentWallet.isApproved ? 'APPROVED' : 'UNKNOWN'
        console.log(`Agent wallet status: ${status}`)
        if (status !== 'ACTIVE' && status !== 'APPROVED' && !agentWallet.isApproved) {
          console.warn(`⚠️ Agent wallet may not be approved. Status: ${status}`)
        }
      }

      // Build payload - check if vaultAddress is needed from agent wallet
      const agentAddress = agentWallet?.address || agentWallet?.agentAddress
      
      // If agent wallet is empty, try a simpler long-only trade first
      // This helps diagnose if the issue is with basket complexity or agent wallet
      const useSimpleTrade = isEmptyWallet && longAssetsFormatted.length > 0
      
      const payload: any = {
        executionType: 'MARKET', // or 'TWAP' for time-weighted
        leverage: 1, // Default 1x leverage
        usdValue: notional, // Must be a number, not a string, and >= 1
        slippage: 0.05, // Between 0.001 and 0.1 (5% default)
        longAssets: useSimpleTrade 
          ? [longAssetsFormatted[0]] // Try just the first asset if agent wallet is empty
          : longAssetsFormatted,
        shortAssets: useSimpleTrade 
          ? [] // No shorts for simple test
          : shortAssetsFormatted,
        // Builder code is included automatically by Pear Protocol if approved
      }
      
      if (useSimpleTrade) {
        console.log('⚠️ Using simplified trade (long-only, single asset) due to empty agent wallet')
        console.log('⚠️ This is a diagnostic test. Full basket trades require agent wallet approval.')
      }
      
      // Add agent address if available (may be required for some endpoints)
      if (agentAddress) {
        console.log(`Using agent address: ${agentAddress}`)
        // Note: Some endpoints may require agentAddress or vaultAddress
        // Uncomment if needed: payload.agentAddress = agentAddress
      }

      // Use the correct endpoint from Pear Protocol API documentation
      // POST /positions is the correct endpoint for basket trades
      const endpoint = '/positions'
      
      console.log(`Executing basket trade on endpoint: ${endpoint}`)
      console.log('Trade payload:', JSON.stringify(payload, null, 2))
      
      try {
        const response = await this.client.post(endpoint, payload)
        console.log('Trade response:', JSON.stringify(response.data, null, 2))
        return this._parseTradeResponse(response.data)
      } catch (error: any) {
        // Enhanced error logging for debugging
        if (axios.isAxiosError(error)) {
          console.error('Trade execution error:', {
            status: error.response?.status,
            statusText: error.response?.statusText,
            data: error.response?.data,
            requestPayload: payload, // Include payload in error log
          })
        }
        throw error
      }
    } catch (error) {
      // Handle Axios errors
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError
        const errorMessage = axiosError.response?.data
          ? JSON.stringify(axiosError.response.data)
          : axiosError.message

        console.error('Pear Protocol API Error:', {
          status: axiosError.response?.status,
          statusText: axiosError.response?.statusText,
          data: axiosError.response?.data,
        })

        throw new Error(
          `Pear Protocol API error: ${axiosError.response?.status || 'Unknown'} - ${errorMessage}`
        )
      }

      // Handle other errors
      throw error instanceof Error
        ? error
        : new Error('Unknown error executing basket trade')
    }
  }
}

// Export a singleton instance
// Note: Instantiated at module load time, which is fine for API routes
// The error.tsx and global-error.tsx files handle any initialization errors gracefully
export const pearClient = new PearClient()

File: test/pear_client.py
"""
Pear Protocol API Client
A Python client for integrating with Pear Protocol's pair trading API on Hyperliquid.
"""

import json
import time
import requests
from typing import Optional, Dict, Any
from eth_account import Account
from eth_account.messages import encode_typed_data
from web3 import Web3


class PearProtocolClient:
    """Client for interacting with Pear Protocol API."""
    
    BASE_URL = "https://hl-v2.pearprotocol.io"
    BUILDER_ADDRESS = "0xA47D4d99191db54A4829cdf3de2417E527c3b042"
    
    def __init__(self, wallet_private_key: Optional[str] = None, wallet_address: Optional[str] = None):
        """
        Initialize the Pear Protocol client.
        
        Args:
            wallet_private_key: Private key of the wallet (hex string with or without 0x prefix)
            wallet_address: Wallet address (will be derived from private key if not provided)
        """
        self.access_token: Optional[str] = None
        self.refresh_token: Optional[str] = None
        self.wallet_address = wallet_address
        self.private_key = wallet_private_key
        
        if wallet_private_key:
            # Ensure private key has 0x prefix
            if not wallet_private_key.startswith('0x'):
                wallet_private_key = '0x' + wallet_private_key
            
            self.account = Account.from_key(wallet_private_key)
            if not self.wallet_address:
                self.wallet_address = self.account.address
        else:
            self.account = None
    
    def _get_headers(self, include_auth: bool = True) -> Dict[str, str]:
        """Get request headers with optional authorization."""
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json"
        }
        if include_auth and self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"
        return headers
    
    def get_eip712_message(self, client_id: str) -> Dict[str, Any]:
        """
        Get EIP-712 message to sign for authentication.
        
        Args:
            client_id: Client ID (e.g., 'APITRADER' for individual traders)
        
        Returns:
            Response containing the EIP-712 message to sign
        """
        if not self.wallet_address:
            raise ValueError("Wallet address is required to get EIP-712 message")
        
        url = f"{self.BASE_URL}/auth/eip712-message"
        params = {
            "address": self.wallet_address,
            "clientId": client_id
        }
        
        response = requests.get(url, params=params, headers=self._get_headers(include_auth=False))
        response.raise_for_status()
        return response.json()
    
    def sign_eip712_message(self, eip712_message: Dict[str, Any]) -> str:
        """
        Sign EIP-712 message using wallet private key.
        
        Args:
            eip712_message: The EIP-712 message structure from get_eip712_message
        
        Returns:
            Signature string (hex with 0x prefix)
        """
        if not self.account:
            raise ValueError("Wallet private key is required for signing")
        
        # Try encoding with full_message first (most common approach)
        try:
            encoded_message = encode_typed_data(full_message=eip712_message)
        except (TypeError, KeyError) as e:
            # Fallback: if full_message doesn't work, try explicit parameters
            # Some APIs return the message in a different structure
            if "domain" in eip712_message and "types" in eip712_message and "message" in eip712_message:
                encoded_message = encode_typed_data(
                    domain_data=eip712_message["domain"],
                    message_types=eip712_message["types"],
                    message_data=eip712_message["message"]
                )
            else:
                raise ValueError(f"Cannot encode EIP-712 message: {e}. Message structure: {list(eip712_message.keys())}")
        
        # Sign the encoded message
        signed_message = self.account.sign_message(encoded_message)
        return signed_message.signature.hex()
    
    def authenticate(self, client_id: str, signature: str, eip712_message: Optional[Dict[str, Any]] = None, timestamp: Optional[str] = None) -> Dict[str, Any]:
        """
        Authenticate with Pear Protocol using signed EIP-712 message.
        
        Args:
            client_id: Client ID used for authentication
            signature: The signed EIP-712 message signature
            eip712_message: Optional EIP-712 message (some APIs require this in details)
            timestamp: Optional timestamp from EIP-712 message
        
        Returns:
            Response containing access_token and refresh_token
        """
        if not self.wallet_address:
            raise ValueError("Wallet address is required for authentication")
        
        url = f"{self.BASE_URL}/auth/login"
        
        # Ensure signature has 0x prefix (standard format for EIP-712)
        signature_formatted = signature if signature.startswith('0x') else '0x' + signature
        
        # Use the same address format as used when requesting the EIP-712 message
        payload = {
            "method": "eip712",
            "address": self.wallet_address,
            "clientId": client_id,
            "details": {
                "signature": signature_formatted
            }
        }
        
        # Add timestamp if provided
        if timestamp:
            payload["details"]["timestamp"] = timestamp
        
        # Note: Not including full message in details as most APIs just need signature
        # If this fails, we may need to try including the message
        
        response = requests.post(url, json=payload, headers=self._get_headers(include_auth=False))
        response.raise_for_status()
        data = response.json()
        
        # Store tokens (handle both camelCase and snake_case)
        self.access_token = data.get("access_token") or data.get("accessToken")
        self.refresh_token = data.get("refresh_token") or data.get("refreshToken")
        
        return data
    
    def login(self, client_id: str) -> Dict[str, Any]:
        """
        Complete login flow: get message, sign, and authenticate.
        
        Args:
            client_id: Client ID to use for authentication
        
        Returns:
            Authentication response with tokens
        """
        if not self.account:
            raise ValueError("Wallet private key is required for login")
        
        # Step 1: Get EIP-712 message
        print(f"[1/3] Getting EIP-712 message for client ID: {client_id}")
        eip712_message = self.get_eip712_message(client_id)
        
        # Extract timestamp if present in message
        timestamp = eip712_message.get("message", {}).get("timestamp")
        
        # Step 2: Sign the message
        print(f"[2/3] Signing EIP-712 message...")
        signature = self.sign_eip712_message(eip712_message)
        
        # Step 3: Authenticate
        print(f"[3/3] Authenticating...")
        auth_response = self.authenticate(client_id, signature, eip712_message, timestamp)
        
        print(f"✓ Authentication successful!")
        print(f"  Access token expires in: 15 minutes")
        print(f"  Refresh token expires in: 30 days")
        
        return auth_response
    
    def refresh_access_token(self) -> Dict[str, Any]:
        """
        Refresh the access token using the refresh token.
        
        Returns:
            New authentication response with tokens
        """
        if not self.refresh_token:
            raise ValueError("No refresh token available. Please authenticate first.")
        
        url = f"{self.BASE_URL}/auth/refresh"
        payload = {"refreshToken": self.refresh_token}
        
        response = requests.post(url, json=payload, headers=self._get_headers(include_auth=False))
        response.raise_for_status()
        data = response.json()
        
        # Update tokens (handle both camelCase and snake_case)
        self.access_token = data.get("access_token") or data.get("accessToken")
        self.refresh_token = data.get("refresh_token") or data.get("refreshToken") or self.refresh_token
        
        return data
    
    def logout(self) -> Dict[str, Any]:
        """
        Logout and invalidate refresh token.
        
        Returns:
            Logout response
        """
        if not self.refresh_token:
            print("No refresh token to invalidate")
            return {}
        
        url = f"{self.BASE_URL}/auth/logout"
        payload = {"refreshToken": self.refresh_token}
        
        response = requests.post(url, json=payload, headers=self._get_headers(include_auth=False))
        response.raise_for_status()
        
        # Clear tokens
        self.access_token = None
        self.refresh_token = None
        
        return response.json()
    
    def get_agent_wallet(self) -> Dict[str, Any]:
        """
        Get agent wallet information.
        
        Returns:
            Agent wallet details including address and status
        """
        url = f"{self.BASE_URL}/agentWallet"
        
        response = requests.get(url, headers=self._get_headers())
        response.raise_for_status()
        return response.json()
    
    def create_agent_wallet(self) -> Dict[str, Any]:
        """
        Create a new agent wallet.
        
        Returns:
            New agent wallet details
        """
        url = f"{self.BASE_URL}/agentWallet"
        
        response = requests.post(url, headers=self._get_headers())
        response.raise_for_status()
        return response.json()
    
    def get_builder_code(self) -> Dict[str, str]:
        """
        Get builder code information.
        
        Returns:
            Builder code address
        """
        return {
            "builder_address": self.BUILDER_ADDRESS,
            "note": "All trades are routed to this builder address. Users must approve this address to charge fees."
        }
    
    def test_connection(self) -> bool:
        """
        Test basic API connection (no authentication required).
        
        Returns:
            True if connection successful
        """
        try:
            # Try to get EIP-712 message as a connection test
            # Using a dummy request to check if API is reachable
            response = requests.get(self.BASE_URL, timeout=5)
            return True
        except Exception as e:
            print(f"Connection test failed: {e}")
            return False

